mlabs/src/Mlabs/NFT/Validation.hs

```diff
@@ -459,10 +475,16 @@ getInputDatums ctx =
 
 -- | Retuns datums attached to outputs of transaction
 getOutputDatums :: PlutusTx.FromData a => ScriptContext -> [a]
-getOutputDatums ctx =
-  mapMaybe (PlutusTx.fromBuiltinData . getDatum)
-    . mapMaybe (\hash -> findDatum hash $ scriptContextTxInfo ctx)
-    . mapMaybe txOutDatumHash
+getOutputDatums = fmap fst . getOutputDatumsWithTx
+
+{-# INLINEABLE getOutputDatumsWithTx #-}
+
+-- | Returns datums and coresponding UTXOs attached to outputs of transaction
+getOutputDatumsWithTx :: PlutusTx.FromData a => ScriptContext -> [(a, TxOut)]
```

Comment:
This is a much better way. Can we implement a similar utility/check for the inputs?